incuna_auth/middleware/permission_feincms.py

```diff
@@ -0,0 +1,74 @@
+from .permission import BasePermissionMiddleware
+from ..models import AccessStateExtensionMixin as AccessState
+
+
+class FeinCMSPermissionMiddleware(BasePermissionMiddleware):
+    """
+    Middleware that allows or denies access based on the resource's access state.
+
+    Access state is provided by incuna_auth.models.AccessStateExtensionMixin and is used
+    as a marker to determine how a FeinCMS resource should be access-controlled.
+
+    The reasoning behind this is twofold. FeinCMS resources have unpredictable URLs, so
+    protecting them with the URL-based middleware is risky. Additionally, a client might
+    want to add a variety of permissions on their Pages (or similar) that don't
+    necessarily correspond to their location within the site. Adding an access_state
+    and checking that through middleware allows maximum flexibility and puts permissions
+    on the page level fully within the user's control.
+
+    This class protects all pages with an access_state of STATE_AUTH_ONLY. To protect
+    a different state or list of states, override get_protected_states.
+    """
+    def get_protected_states(self):
+        """
+        Returns a list of access states this middleware should apply to.
+
+        By default, returns STATE_AUTH_ONLY, which is the only non-custom access state
+        that implies any restriction.
+        """
+        return [AccessState.STATE_AUTH_ONLY]
+
+    def _get_page_from_path(self, path):
+        """
+        Fetches the FeinCMS Page object that the path points to.
+
+        Override this to deal with different types of object from Page.
+        """
+        from feincms.module.page.models import Page
```

Comment:
Using `Page.objects.best_match_for_path()` requires a lot of mocking in tests.  The import for FeinCMS's `Page` is inside the `_get_path_for_page()` method in the middleware class, so that we can avoid importing it in the tests.